cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -83,40 +87,151 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
+                }, true,
+                res -> {
+                    if (res.succeeded()) {
+                        fut.complete((KafkaClusterDescription) res.result());
+                    } else {
+                        fut.fail("");
+                    }
+                }
+        );
+        return fut;
+    }
+
+    private final Future<Void> createOrUpdateKafka(Reconciliation reconciliation, ConfigMap assemblyCm, List<Secret> assemblySecrets) {
         String namespace = assemblyCm.getMetadata().getNamespace();
         String name = assemblyCm.getMetadata().getName();
         log.debug("{}: create/update kafka {}", reconciliation, name);
-        KafkaCluster kafka;
-        try {
-            kafka = KafkaCluster.fromConfigMap(assemblyCm);
-        } catch (Exception e) {
-            return Future.failedFuture(e);
-        }
-        Service service = kafka.generateService();
-        Service headlessService = kafka.generateHeadlessService();
-        ConfigMap metricsConfigMap = kafka.generateMetricsConfigMap();
-        StatefulSet statefulSet = kafka.generateStatefulSet(isOpenShift);
 
         Future<Void> chainFuture = Future.future();
-        kafkaSetOperations.scaleDown(namespace, kafka.getName(), kafka.getReplicas())
-                .compose(scale -> serviceOperations.reconcile(namespace, kafka.getName(), service))
-                .compose(i -> serviceOperations.reconcile(namespace, kafka.getHeadlessName(), headlessService))
-                .compose(i -> configMapOperations.reconcile(namespace, kafka.getMetricsConfigName(), metricsConfigMap))
-                .compose(i -> kafkaSetOperations.reconcile(namespace, kafka.getName(), statefulSet))
-                .compose(diffs -> kafkaSetOperations.maybeRollingUpdate(diffs.resource()))
-                .compose(i -> kafkaSetOperations.scaleUp(namespace, kafka.getName(), kafka.getReplicas()))
-                .compose(scale -> serviceOperations.endpointReadiness(namespace, service, 1_000, operationTimeoutMs))
-                .compose(i -> serviceOperations.endpointReadiness(namespace, headlessService, 1_000, operationTimeoutMs))
-                .compose(chainFuture::complete, chainFuture);
+        getKafkaCluster(assemblyCm, assemblySecrets)
+                .compose(desc -> desc.withVoid(kafkaSetOperations.scaleDown(namespace, desc.kafka().getName(), desc.kafka().getReplicas())))
+                .compose(desc -> desc.withVoid(serviceOperations.reconcile(namespace, desc.kafka().getName(), desc.service())))
+                .compose(desc -> desc.withVoid(serviceOperations.reconcile(namespace, desc.kafka().getHeadlessName(), desc.headlessService())))
+                .compose(desc -> desc.withVoid(configMapOperations.reconcile(namespace, desc.kafka().getMetricsConfigName(), desc.metricsConfigMap())))
+                .compose(desc -> desc.withVoid(secretOperations.reconcile(namespace, desc.kafka().getName() + "-clients-ca", desc.clientsCASecret())))
```

Comment:
The name suffixes of the secrets are all over the place. I think you should have them defined centrally and not on 10 different places.

Reply:
Yeah :-)